src/mpid/ch4/include/mpidpre.h

```diff
@@ -227,45 +227,57 @@ struct MPIDI_CH4U_win_lock {
     uint16_t type;
 };
 
-struct MPIDI_CH4U_win_queue {
+typedef struct MPIDI_CH4U_win_lock_recvd {
     struct MPIDI_CH4U_win_lock *head;
     struct MPIDI_CH4U_win_lock *tail;
-};
+    int type;                   /* current lock's type */
+    unsigned count;             /* count of granted locks (not received) */
+} MPIDI_CH4U_win_lock_recvd_t;
 
-typedef struct MPIDI_CH4U_win_lock_info {
-    unsigned peer;
-    int lock_type;
-    struct MPIR_Win *win;
-    volatile unsigned done;
-} MPIDI_CH4U_win_lock_info;
+typedef struct MPIDI_CH4U_win_target_sync_lock {
+    volatile unsigned locked;   /* locked == 0 or 1 */
```

Comment:
I would think the order of lock and unlock in your example should be the other way around, but anyway.

Same discussion for reordering applies to the register caching as well. Since `pthread_mutex_lock/unlock` is supposed to synchronize memory with respect to other threads, compiler cannot cache values (that may be shared with other threads) in the register if these calls appear.

>The compiler (just tried GCC) does not do the register caching or replacing with constant for cond, is not because of the barrier thing, it is actually because it cannot understand the behavior of external functions such as pthread_mutex_lock.

This is also an important point. Beside the specification of `pthread_mutex_lock` above, lock primitives are most likely implemented as external library functions, so compilers cannot cache values anyway.

To summarize, I strongly believe we don't need `volatile` as long as pthread mutex is used.